### ISSUE TITLE
bad mapmatching results when oneway=false and reversed=true

### ISSUE THREAD

**Issue by wharstr0927** [gboeing/osmnx#1346]
### Contributing guidelines

- [x] I understand the contributing guidelines

### Documentation

- [x] My problem is not addressed by the documentation or examples

### Existing issues

- [x] My problem does not appear in an existing issue

### What operating system and Python version are you using?

linux,Python 3.8.10 

### What OSMnx version are you using?

2.0.6

### Environment packages and versions

```shell
root@0660f4a5c0dc:/mnt/workplace/online_mapmatch# pip list
Package             Version             
------------------- --------------------
asttokens           3.0.0               
attrs               25.1.0              
backcall            0.2.0               
certifi             2019.11.28          
chardet             3.0.4               
click               8.1.8               
click-plugins       1.1.1               
cligj               0.7.2               
comm                0.2.2               
contourpy           1.1.1               
cycler              0.12.1              
cykhash             2.0.1               
Cython              3.0.12              
dbus-python         1.2.16              
debugpy             1.8.12              
decorator           5.1.1               
distro-info         0.23+ubuntu1.1      
executing           2.2.0               
fiona               1.9.6               
fonttools           4.56.0              
GDAL                3.3.2               
geopandas           0.10.2              
idna                2.8                 
importlib-metadata  8.5.0               
importlib-resources 6.4.5               
ipykernel           6.29.5              
ipython             8.12.3              
jedi                0.19.2              
jinja2              3.1.5               
joblib              1.4.2               
jupyter-client      8.6.3               
jupyter-core        5.7.2               
kiwisolver          1.4.7               
lxml                5.3.1               
MarkupSafe          2.1.5               
matplotlib          3.7.5               
matplotlib-inline   0.1.7               
nest-asyncio        1.6.0               
numpy               1.24.4              
osmium              4.0.2               
packaging           24.2                
pandas              2.0.3               
parso               0.8.4               
pexpect             4.9.0               
pickleshare         0.7.5               
pillow              10.4.0              
pip                 20.0.2              
platformdirs        4.3.6               
prompt-toolkit      3.0.50              
psutil              7.0.0               
ptyprocess          0.7.0               
pure-eval           0.2.3               
pygments            2.19.1              
PyGObject           3.36.0              
pyparsing           3.1.4               
pyproj              3.5.0               
pyrobuf             0.9.3               
pyrosm              0.6.1               
python-apt          2.0.1+ubuntu0.20.4.1
python-dateutil     2.9.0.post0         
python-rapidjson    1.20                
pytz                2025.1              
pyzmq               26.2.1              
requests            2.22.0              
requests-unixsocket 0.2.0               
scikit-learn        1.3.2               
scipy               1.10.1              
setuptools          45.2.0              
shapely             2.0.3               
six                 1.14.0              
stack-data          0.6.3               
threadpoolctl       3.5.0               
tornado             6.4.2               
tqdm                4.67.1              
traitlets           5.14.3              
typing-extensions   4.12.2              
tzdata              2025.1              
unattended-upgrades 0.1                 
unzip               1.0.0               
urllib3             1.25.8              
wcwidth             0.2.13              
wheel               0.34.2              
zipp                3.20.2
```

### How did you install OSMnx?

OSMnx Docker image

### Problem description

I keep trying many trials about the mapmatching process, and per step operation, but so far so bad, I am exhausted, so I have to ask for help, thank you so much if I can get any response. 
As you see I have four <u,v> pairs, but no u in the ubodt file as source id, I am so confused. 
any comments I will appreciate.
<img width="582" height="201" alt="Image" src="https://github.com/user-attachments/assets/2127fe32-5e8f-4f03-a245-23566a735627" />

<img width="540" height="311" alt="Image" src="https://github.com/user-attachments/assets/b2411b04-f749-48cd-9276-d8a09a169687" />

### Minimal standalone reproducible example

```python
#!/usr/bin/env python
# coding: utf-8

# encoding = "utf-8"
import os
import osmnx as ox
import numpy as np
import pandas as pd
from shapely.geometry import LineString
encoding = "utf-8"


city_names=['北京市']
for city in city_names:
    cmd = f"osmium cat beijing-latest.osm.pbf --overwrite -o ./{city}-latest.osm"
    print(f"执行：{cmd}")
    get_ipython().system('{cmd}')


#debug
city_name = "beijingshi"
# encoding = "utf-8"

print(f"==> 正在处理：{city_name}")
osm_path = f"./北京市-latest.osm"
output_dir = f"./{city_name}"
os.makedirs(output_dir, exist_ok=True)


G = ox.graph_from_xml(osm_path, bidirectional=False, simplify=False, retain_all=True)
# G = ox.graph_from_xml(osm_path, bidirectional=False, simplify=True, retain_all=True)

gdf_edges, gdf_edges = ox.graph_to_gdfs(G)
print(f"原始道路条数：{len(gdf_edges)}")


gdf_nodes = gdf_nodes.reset_index()    # 如果你也想把节点的 osmid 变成一列
gdf_edges = gdf_edges.reset_index()    # u, v, key → 普通列
# gdf_edges.head()
# ——— 3. 非数值列字符串化，避免写 Shapefile 时出错 ———
gdf_edges = ox.io._stringify_nonnumeric_cols(gdf_edges)
len(gdf_edges)
# gdf_edges

gdf_nodes[gdf_nodes['osmid'] == 1072576082]

mask_highway = gdf_edges['highway'].notnull() & (gdf_edges['highway'] != '')
gdf_edges = gdf_edges[mask_highway].copy()

# 6. 重建 fid
gdf_edges['fid'] = np.arange(len(gdf_edges), dtype=int)


gdf_edges[gdf_edges['osmid']==201724128][['u','v','osmid','oneway','reversed', 'fid', 'geometry']]

gdf_edges[gdf_edges['osmid']=='1072576086'][['u','v','osmid','oneway','reversed', 'fid', 'geometry']]


gdf_edges[gdf_edges['osmid']=='1072576082'][['u','v','osmid','oneway','reversed', 'fid', 'geometry']]


# 7. 输出
edges_path = os.path.join(output_dir, "edges.shp")
gdf_edges.to_file(edges_path, encoding=encoding)
print(f"生成{city_name}shp文件")
print(f"→ 已保存边文件（含过滤+双向+fid），共 {len(gdf_edges)} 条：{edges_path}")
```

**Comment by wharstr0927** [comment]
Actually,I installed it using pip, sorry I made a mistake, real pip list as follows:Package                   Version
------------------------- --------------
absl-py                   2.2.1
annotated-types           0.7.0
antlr4-python3-runtime    4.9.3
anyio                     4.9.0
argon2-cffi               23.1.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
asttokens                 3.0.0
async-lru                 2.0.5
attrs                     25.3.0
babel                     2.17.0
beautifulsoup4            4.13.4
bleach                    6.2.0
branca                    0.8.2
certifi                   2024.12.14
cffi                      1.17.1
charset-normalizer        3.4.1
click                     8.1.8
comm                      0.2.1
contourpy                 1.3.2
cycler                    0.12.1
datasets                  2.17.1
debugpy                   1.8.14
decorator                 5.1.1
defusedxml                0.7.1
descartes                 1.1.0
dill                      0.3.8
docker-pycreds            0.4.0
einops                    0.7.0
einshape                  1.0
executing                 2.1.0
fastjsonschema            2.21.1
filelock                  3.16.1
folium                    0.20.0
fonttools                 4.58.1
fqdn                      1.5.1
fsspec                    2024.12.0
geopandas                 1.1.1
gitdb                     4.0.12
GitPython                 3.1.44
gluonts                   0.14.4
grpcio                    1.71.0
h11                       0.16.0
httpcore                  1.0.9
httpx                     0.28.1
huggingface-hub           0.29.3
hydra-core                1.3.0
idna                      3.10
inquirerpy                0.3.4
ipykernel                 6.29.5
ipython                   8.31.0
ipython_pygments_lexers   1.1.1
isoduration               20.11.0
jax                       0.5.3
jax-cuda12-pjrt           0.4.38
jax-cuda12-plugin         0.4.38
jaxlib                    0.5.3
jaxtyping                 0.2.38
jedi                      0.19.2
Jinja2                    3.1.5
joblib                    1.4.2
json5                     0.12.0
jsonpointer               3.0.0
jsonschema                4.24.0
jsonschema-specifications 2025.4.1
jupyter_client            8.6.3
jupyter_core              5.8.1
jupyter-events            0.12.0
jupyter-lsp               2.2.5
jupyter_server            2.16.0
jupyter_server_terminals  0.5.3
jupyterlab                4.4.3
jupyterlab_pygments       0.3.0
jupyterlab_server         2.27.3
kiwisolver                1.4.8
lightning                 2.5.1
lightning-utilities       0.14.2
markdown-it-py            3.0.0
MarkupSafe                3.0.2
matplotlib                3.10.3
matplotlib-inline         0.1.6
mdurl                     0.1.2
mistune                   3.1.3
ml_dtypes                 0.5.1
mpmath                    1.3.0
multiprocess              0.70.16
nbclient                  0.10.2
nbconvert                 7.16.6
nbformat                  5.10.4
nest-asyncio              1.6.0
networkx                  3.4.2
notebook                  7.4.3
notebook_shim             0.2.4
numpy                     2.2.1
nvidia-cublas-cu12        12.1.3.1
nvidia-cuda-cupti-cu12    12.1.105
nvidia-cuda-nvcc-cu12     12.6.85
nvidia-cuda-nvrtc-cu12    12.1.105
nvidia-cuda-runtime-cu12  12.1.105
nvidia-cudnn-cu12         9.1.0.70
nvidia-cufft-cu12         11.0.2.54
nvidia-curand-cu12        10.3.2.106
nvidia-cusolver-cu12      11.4.5.107
nvidia-cusparse-cu12      12.1.0.106
nvidia-nccl-cu12          2.20.5
nvidia-nvjitlink-cu12     12.8.93
nvidia-nvtx-cu12          12.1.105
omegaconf                 2.3.0
opt_einsum                3.4.0
orjson                    3.10.16
osmnx                     2.0.6
overrides                 7.7.0
packaging                 24.2
pandas                    2.2.3
pandocfilters             1.5.1
parso                     0.8.4
pexpect                   4.9.0
pfzy                      0.3.4
pillow                    11.2.1
pip                       24.2
platformdirs              4.3.6
prometheus_client         0.22.0
prompt_toolkit            3.0.48
protobuf                  5.29.3
psutil                    5.9.0
ptyprocess                0.7.0
pure-eval                 0.2.2
pyarrow-hotfix            0.6
pycparser                 2.22
pydantic                  2.10.5
pydantic_core             2.27.2
Pygments                  2.19.1
pyogrio                   0.11.1
pyparsing                 3.2.3
pyproj                    3.7.2
python-dateutil           2.9.0.post0
python-dotenv             1.0.0
python-json-logger        3.3.0
pytorch-lightning         2.5.1
pytz                      2024.2
PyYAML                    6.0.2
pyzmq                     26.2.0
referencing               0.36.2
requests                  2.32.3
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      13.9.4
rpds-py                   0.25.1
rtree                     1.4.1
safetensors               0.5.3
scikit-learn              1.6.0
scipy                     1.15.0
Send2Trash                1.8.3
sentry-sdk                2.19.2
setproctitle              1.3.4
setuptools                75.1.0
shapely                   2.1.2
shellingham               1.5.4
six                       1.17.0
smmap                     5.0.2
sniffio                   1.3.1
soupsieve                 2.7
stack-data                0.2.0
sympy                     1.13.1
tensorboard               2.19.0
tensorboard-data-server   0.7.2
terminado                 0.18.1
threadpoolctl             3.5.0
tinycss2                  1.4.0
torch                     2.4.1
torchmetrics              1.7.0
tornado                   6.5.1
tqdm                      4.67.1
traitlets                 5.14.3
triton                    3.0.0
typer                     0.15.1
types-python-dateutil     2.9.0.20250516
typing_extensions         4.12.2
tzdata                    2024.2
uri-template              1.3.0
urllib3                   2.3.0
utilsforecast             0.2.10
wadler_lindig             0.1.4
wandb                     0.19.2
wcwidth                   0.2.13
webcolors                 24.11.1
webencodings              0.5.1
websocket-client          1.8.0
wheel                     0.44.0
xxhash                    3.5.0
xyzservices               2025.4.0

**Comment by wharstr0927** [comment]
> Please provide a _complete_ and _minimal_ reproducible example. Complete means it doesn't rely on any external files or definitions beyond what's in the code snippet. Minimal means there's nothing in the code snippet beyond the minimum necessary to demonstrate the specific problem. Then I can take a look and troubleshoot.

ok, I will do it. thank you.



[test_min_example.py](https://github.com/user-attachments/files/23573215/test_min_example.py)
